R/mod-button-indicator.R

```diff
@@ -0,0 +1,108 @@
+#' These functions add button feedback features including:
+#' disabling the button while processing requested function,
+#' showing a spinning wheel while processing requested function,
+#' displaying a green checkmark showing success upon completion,
+#' or displaying an error message if the function requested failed.
+#'
+#' Author: Dean Attali
+#' From https://github.com/daattali/advanced-shiny/blob/master/busy-indicator/helpers.R #nolint
+#'
+#' Hint for making this work with modules by mmoise in PR#11
+
+#' Wrap the button in this function to attach visual features
+#' @param button a shiny actionButton
+with_busy_indicator_ui <- function(button) {
+  id <- button[["attribs"]][["id"]]
+  div(
+    shinyjs::useShinyjs(),
+    `data-for-btn` = id,
+    button,
+    span(
+      class = "btn-loading-container",
+      shinyjs::hidden(
+        icon("spinner", class = "btn-loading-indicator fa-spin"),
+        icon("check", class = "btn-done-indicator")
+      )
+    ),
+    shinyjs::hidden(
+      div(
+        class = "btn-err",
+        div(
+          icon("exclamation-circle"),
+          tags$b("Error: "),
+          span(class = "btn-err-msg")
+        )
+      )
+    )
+  )
+}
+
+#' When a button is clicked, call this function from the server.
+#' This will execute the expression desired and signal the visual
+#' feedback for the button. If the expression fails, the error
+#' message displayed is from the error that was thrown.
+#' @param button_id id of shiny actionButton
+#' @param expr the code to run when the button is clicked
+with_busy_indicator_server <- function(button_id, expr) {
+  # UX stuff: show the "busy" message, hide the other messages,
+  # disable the button
+  # Need to get session in order for button to indicate correctly
+  session <- shinyjs:::getSession()
```

Comment:
We should try to avoid relying on unexported functions if possible (otherwise R CMD check will complain). It looks like `shinyjs:::getSession()` is basically a wrapper for `shiny::getDefaultReactiveDomain()` which is exported from shiny; do you know if it would work to use the latter instead?

Reply:
This does work! Fixed.